talk/expert/cpp20spaceship.tex

```diff
@@ -0,0 +1,242 @@
+      \item I am not aware of some predefined type whose {\it <=>} would return an instance of {\it std::weak_ordering}, but it may make sense for some home-made class, such as the following.
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Home-made class}
+    \begin{itemize}
+    \item In the example below, we define a very basic class for positive rational numbers, and provide an implementation of {\it <=>}.
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    \begin{cppcode*}{}
+#include <iostream>
+#include <compare>
+
+struct Rational
+ {
+  unsigned n, d ;
+
+  std::weak_ordering operator<=>( Rational const & other ) const
+   { return (n*other.d)<=>(other.n*d) ; }
+ } ;
+
+std::ostream & operator<<( std::ostream & os, std::weak_ordering cmp )
+ { return (os<<'<'<<(cmp<0)<<'|'<<(cmp==0)<<'|'<<(cmp>0)<<'>') ; }
+
+std::ostream & operator<<( std::ostream & os, Rational const & r )
+ { return (os<<r.n<<'/'<<r.d) ; }
+
+template <typename T>
+void compare( T lhs, T rhs )
+ {
+  std::cout<<std::endl ;
+  //std::cout<<lhs<<" ==  "<<rhs<<": "<<(lhs==rhs)<<std::endl ;
+  //std::cout<<lhs<<" !=  "<<rhs<<": "<<(lhs!=rhs)<<std::endl ;
+  std::cout<<lhs<<" <=> "<<rhs<<": "<<(lhs<=>rhs)<<std::endl ;
+  std::cout<<lhs<<" <   "<<rhs<<": "<<(lhs<  rhs)<<std::endl ;
+  std::cout<<lhs<<" <=  "<<rhs<<": "<<(lhs<= rhs)<<std::endl ;
+  std::cout<<lhs<<" >   "<<rhs<<": "<<(lhs>  rhs)<<std::endl ;
+  std::cout<<lhs<<" >=  "<<rhs<<": "<<(lhs>= rhs)<<std::endl ;
+ }
+
+int main()
+ {
+  compare<Rational>({ 3, 4 },{ 2, 3 }) ;
+  compare<Rational>({ 3, 6 },{ 2, 3 }) ;
+  compare<Rational>({ 1, 2 },{ 2, 4 }) ;
+ }
+     \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Home-made class}
+    \begin{itemize}
+      \item Despites the {\it <=>} for {\it unsigned} returns some instance of {\it std::strong_ordering}, we prefer here to cast it to {\it std::weak_ordering}. This way, we emphasize that if {\it a<=>b} is equal to {\it 0}, it only means that {\it a} and {\it b} are logically equivalent, but may lead to different results in some expressions.
+      \item As one can check in the code, we have defined {\it <=>}, and the compiler has added, for free, some default implementations for {\it <}, {\it >}, {\it <=} and {\it =>}. Of course, one can overload them with home-made implementations.
+      \item It has NOT added a default implementation for {\it ==} and {\it !=}. Those operators are generally expected to mean \textbf{equal}, rather than \textbf{equivalent}. If {\it <=>} does not enable a strong order, the operator {\it ==}generally deserves a separate implementation.
```

Comment:
I disagree. `operator==` is used to express the same ordering relation as `operator<=>`. So if your class is weakly ordered, then `operator==` expresses equivalence and not equality.
```suggestion
      \item The compiler will NOT express {\it ==} and {\it !=} in terms of {\it <=>}. If your class implements {\it <=>}, it should also implement {\it ==}. The compiler can however express {\it !=} in terms of {\it ==}, so a separate implementation of {\it !=} is no longer required.
```

Reply:
I would be interested by some resources/references about this point. You describe the class has having a class-level kind of global ordering. If so, I do not understand why `==` is not deducable from `<=>`, just like other ordering operators. If performance is the reason, we may have a deduction by default, and the possibility to overload `==` when one needs performance, just the same as other operators.

Reply:
@bernhardmgruber Here : https://www.foonathan.net/2018/06/equivalence-relations/, Jonathan Muller explains why one should never implement == if it is not a real strong equality. Perhaps we may agree than when <=> is not a strong ordering, it is adviced to NOT define ==, and when <=> is a strong ordering, == should be defined and equivalent to <=>==0. This would reconcile my feeling that == should mean equality, the fact that the compiler is not allowed to transform == into <=>==0, and your point that == should be equivalent to <=>==0 . What do you think ?

Reply:
> @bernhardmgruber Here : https://www.foonathan.net/2018/06/equivalence-relations/, Jonathan Muller explains why one should never implement == if it is not a real strong equality. Perhaps we may agree than when <=> is not a strong ordering, it is adviced to NOT define ==, and when <=> is a strong ordering, == should be defined and equivalent to <=>==0. This would reconcile my feeling that == should mean equality, the fact that the compiler is not allowed to transform == into <=>==0, and your point that == should be equivalent to <=>==0 . What do you think ?

After reading the linked article, I have come to agree with you! Consider linking to the article on one of your slides!
So `operator==` should be only provided to mean equality, not equivalence. If `operator<=>` specifies a weak/partial ordering, don't define `operator==`. If you still do it for some reason, make it consistent with `operator<=>`.